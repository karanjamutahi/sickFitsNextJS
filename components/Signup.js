import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';

import Form from './styles/Form';
import Error from './ErrorMessage';
import CURRENT_USER_QUERY from './User';

const SIGNUP_MUTATION = gql`
    mutation SIGNUP_MUTATION(
        $email: String!,
        $password: String!,
        $Fname: String!,
        $lastname: String
    ) {
        signUp(
            email: $email
            password: $password
            firstname: $Fname
            lastname: $lastname
        ) {
            id,
            email,
            firstname
        }
    }
`;

class Signup extends Component {
    state = {
        email: '',
        password: '',
        Fname: '',
    };

    saveToState = (e) => {
        this.setState({ [e.target.name] : e.target.value });
    };


    render() {
        return (
            <Mutation mutation={SIGNUP_MUTATION} variables={this.state} refetchQueries={[{query: CURRENT_USER_QUERY}]}>
                {(signup, { loading, error }) => {

                return(
                <Form data-test="signup" method="POST" onSubmit={ async (e) => {
                    e.preventDefault();
                    const res = await signup();
                    this.setState({
                        email: '',
                        password: '',
                        Fname: '',
                    });
                    Router.push('/item');
                }}>
                    <fieldset disabled={loading} aria-busy={loading}>
                        <h2>Sign Up for an Account</h2>
                        <Error error={error}></Error>
                        <label htmlFor="email">
                            Email
                            <input type="email" name="email" id="email" placeholder='jack.doe@gmail.com' required value={this.state.email} onChange={this.saveToState}/>
                        </label>

                        <label htmlFor="Fname">
                            First Name
                            <input type="text" name="Fname" id="Fname" placeholder='Jack' required value={this.state.Fname} onChange={this.saveToState}/>
                        </label>

                        <label htmlFor="password">
                            Password
                            <input type="password" name="password" id="password" placeholder='Enter secure password' required value={this.state.password} onChange={this.saveToState}/>
                        </label>

                        <button type="submit">Sign Up</button>
                    </fieldset>
                </Form>
                )}}
            </Mutation>
        );
    }
}


export default Signup;
export { SIGNUP_MUTATION };