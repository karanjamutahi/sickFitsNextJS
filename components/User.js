import React, { Component } from 'react';
import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';


const CURRENT_USER_QUERY = gql`
    query CURRENT_USER_QUERY {
        me{
            id
            email
            firstname
            orders{
                id
            }
            cart {
                id
                quantity
                item {
                    id
                    price
                    image
                    title
                    description
                }
            }
        }
    }
`;

const User = (props) => (
    <Query query={CURRENT_USER_QUERY} {...props}>
        { 
            (payload) => {
                return props.children(payload);
            } 
        }
    </Query>
);

//Render Prop whose only child can be a function 
User.propTypes = {
    children: PropTypes.func.isRequired,
};
 
export default User;
export { CURRENT_USER_QUERY };