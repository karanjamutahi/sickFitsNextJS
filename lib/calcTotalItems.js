function totalItems(cart) {
    return cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0)
}

export default totalItems;