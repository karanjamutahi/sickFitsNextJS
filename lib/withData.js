import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { devEndpoint, prodEndpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? devEndpoint : devEndpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            //read cartopen value from cache
             const { cartOpen } = cache.readQuery({
               query: LOCAL_STATE_QUERY,
             });
             //Write the cart State to the opposite
             const data = {
               data: {
                 cartOpen: ! cartOpen
               }
             };
             cache.writeData(data);
             return data; 
          }
        }
      },
      defaults: {
        cartOpen: false,
      }
    }
  });
}

export default withApollo(createClient);
