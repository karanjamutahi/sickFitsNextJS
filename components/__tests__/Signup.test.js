import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import Signup, { SIGNUP_MUTATION } from '../Signup';
import { CURRENT_USER_QUERY } from '../User';
import { fakeUser } from '../../lib/testUtils';
import { ApolloConsumer } from 'react-apollo';


function type(wrapper, name, value) {
    //console.log(wrapper.find(`input[name="${name}"]`).debug());
    wrapper.find(`input[name="${name}"]`).simulate('change', {target: { name, value }});
}

const me = fakeUser();
const mocks = [
    {
        request:{ query: SIGNUP_MUTATION, variables: {
            email: me.email,
            password: 'password',
            Fname: me.firstname,
            lastname: 'fanya kitu unataka',
        }},
        result: { data: {
            signUp: {
                __typename: 'User',
                id: 'abc123',
                email: me.email,
                firstname: me.firstname, 
            }
        }}
    },
    {
        request: { query: CURRENT_USER_QUERY,},
        result: { data: { me } },
    }
];

describe("Signup Component", () => {
     it("Renders and matches snapshot", async () => {
        const wrapper = mount(
            <MockedProvider>
                <Signup></Signup>
            </MockedProvider>
        );
        const signup = wrapper.find('form');
        //console.log(toJSON(signup));
        //console.log(wrapper.debug());
        expect(toJSON(signup)).toMatchSnapshot();
     });

     xit('Calls the mutation properly', async () => {
        let apolloClient;
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <ApolloConsumer>
                    { client => {
                        apolloClient = client;
                        return (<Signup></Signup>);
                        } 
                    }
                </ApolloConsumer>
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        //console.log(apolloClient);
        //console.log(wrapper.find('input').debug());
        type(wrapper, 'Fname', me.firstname);
        type(wrapper, 'email', me.email);
        type(wrapper, 'password', "password");
        wrapper.update();
        const form = wrapper.find('form[data-test="signup"]');
        //console.log(form.debug());
        form.simulate('submit');
        await wait();
        const user = await apolloClient.query({ query: CURRENT_USER_QUERY});
        console.log(user);
     });

     
});