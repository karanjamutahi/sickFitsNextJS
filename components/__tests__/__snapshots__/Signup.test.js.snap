// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup Component Renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 gvKJrt"
  data-test="signup"
  method="POST"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Sign Up for an Account
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label
      htmlFor="email"
    >
      Email
      <input
        id="email"
        name="email"
        onChange={[Function]}
        placeholder="jack.doe@gmail.com"
        required={true}
        type="email"
        value=""
      />
    </label>
    <label
      htmlFor="Fname"
    >
      First Name
      <input
        id="Fname"
        name="Fname"
        onChange={[Function]}
        placeholder="Jack"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="password"
    >
      Password
      <input
        id="password"
        name="password"
        onChange={[Function]}
        placeholder="Enter secure password"
        required={true}
        type="password"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Sign Up
    </button>
  </fieldset>
</form>
`;
